onnx/defs/math/defs.cc

```diff
@@ -734,6 +729,76 @@ static const char* MatMul_ver6_doc = R"DOC(
 Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matmul.html
 )DOC";
 
+void matmulShapeInference(InferenceContext& ctx) {
+  if (!hasNInputShapes(ctx, 2)) {
+    return;
+  }
+
+  const auto shape0 = ctx.getInputType(0)->tensor_type().shape();
+  const auto shape1 = ctx.getInputType(1)->tensor_type().shape();
+
+  if (shape0.dim_size() == 0 || shape1.dim_size() == 0) {
+    fail_shape_inference("Input tensors of wrong rank (0).");
+  }
+
+  TensorShapeProto shapeL, shapeR;
+
+  // First promote each shape to at least rank-2. This logic is
+  // specific to matmul, not generic broadcasting.
+  {
+    if (shape0.dim_size() == 1) {
```

Comment:
If shape0.dim_size() == 1... it seems an incorrect case. Either empty or at least rank 2, right?

Reply:
in this case, a dimension with value "1" was added to ensure its rank is greater than 2. This is the old logic (I didn't touch it) implemented by @anderspapitto 

@anderspapitto  may share more details on why we want to support this case.
Thanks!